File: javascript/again/binarySearch.js
//const nums = [1, 2, 3, 4, 5, 6, 7, 8, 9];
//
//const binarySearch = (p, r, n) => {
//    if (p > r) {
//        return -1;
//    }
//
//    const q = (p + r) / 2 >> 0;
//
//    if (nums[q] == n) {
//        return q;
//    }
//
//    if (nums[q] > n) {
//        return binarySearch(p, q - 1, n);
//    } else {
//        return binarySearch(q + 1, r, n);
//    }
//};
//
//console.log(
//    binarySearch(0, nums.length - 1, 9)
//);

const binarySearch = (arr, n) => {
    let min = 0;
    let max = arr.length - 1;

    while (true) {
        const mid = (max + min) / 2 >> 0;

        if (max < min) {
            return -1;
        }

        if (arr[mid] === n) {
            return mid;
        }

        if (arr[mid] > n) {
            max = mid - 1;
        } else {
            min = mid + 1;
        }
    }
};

console.log(
    binarySearch(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
        7
    )
);

File: c/algorithms/binary_search/binary_search.c
#include <stdio.h>

int primes[25] = { 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97 };
unsigned int len = sizeof(primes) / sizeof(len);

int binary_search(int min, int max, int target) {
    if (min > max) {
        return -1;
    }

    int guess = (max + min) / 2;

    if (primes[guess] == target) {
        return target;
    }

    if (primes[guess] < target) {
        return binary_search(guess + 1, max, target);
    } else {
        return binary_search(min, guess - 1, target);
    }
}

void main(void) {
    printf("%d", binary_search(0, len - 1, 97));
}

